src/commands/synthetics/build-and-test.ts

```diff
@@ -1,130 +1,254 @@
+If not, you can learn more about it here: https://github.com/DataDog/build-plugins#readme
+`)
+
+export const MalformedBuildError = new Error(
+  `Invalid payload. Expected payload is {\"outputDirectory\": string, \"publicPath\": string}`
+)
+
+const MIME_TYPES = {
+  default: 'application/octet-stream',
+  html: 'text/html; charset=UTF-8',
+  js: 'application/javascript',
+  css: 'text/css',
+  png: 'image/png',
+  jpg: 'image/jpg',
+  gif: 'image/gif',
+  ico: 'image/x-icon',
+  svg: 'image/svg+xml',
+} as const
+type Extension = keyof typeof MIME_TYPES
+const isMIMEType = (mime: string): mime is Extension => {
+  return Object.keys(MIME_TYPES).includes(mime as Extension)
+}
+
+type File =
   | {
-      readiness: 'buildCommandExited'
+      found: true
+      ext: string
+      content: string
     }
   | {
-      readiness: 'buildCommandErrored'
-      error: Error
+      found: false
     }
 
-export const DEFAULT_BUILD_PLUGIN_PORT = 4000
+const routeVerbs = ['get', 'post', 'put', 'patch', 'delete'] as const
+type RouteVerb = typeof routeVerbs[number]
+const isRouteVerb = (verb: string): verb is RouteVerb => {
+  return routeVerbs.includes(verb as RouteVerb)
+}
 
-export const UnconfiguredBuildPluginError = new Error(`
-We couldn't detect the Datadog Build plugins within your build. Did you add it?
-If not, you can learn more about it here: https://github.com/DataDog/build-plugins#readme
-`)
+export type Routes = Record<
+  string,
+  {
+    [key in RouteVerb]?: (req: http.IncomingMessage, res: http.ServerResponse) => Promise<void>
+  }
+>
+
+export type RequestHandlerOptions = {
+  builds: ReportedBuild[]
+  root?: string
+  routes?: Routes
+}
+
+const fileExists = async (filePath: string): Promise<boolean> => {
+  try {
+    await fs.access(filePath)
+
+    return true
+  } catch {
+    return false
+  }
+}
 
-const watchBuildPluginServerReadiness = async (buildPluginServerUrl: string, abortSignal: AbortSignal) => {
-  return poll(async () => {
-    try {
-      const response = await axios.get<BuildStatus>(buildPluginServerUrl, {signal: abortSignal})
-      const {status, publicPrefix = ''} = response.data
-
-      if (status === 'success') {
-        return {
-          readiness: 'buildCommandReady',
-          publicPrefix,
-        } as const
+const prepareFile = async (root = process.cwd(), builds: ReportedBuild[], requestUrl: string): Promise<File> => {
+  const staticPath = path.isAbsolute(root) ? root : path.resolve(process.cwd(), root)
+
+  for (const build of builds) {
+    const url = new URL(requestUrl, 'http://127.0.0.1')
+    const filePath = path.join(
+      path.resolve(staticPath, build.outputDirectory), // absolute path to the assets directory
+      path.relative(build.publicPath, url.pathname), // relative path to the file
+      url.pathname.endsWith('/') ? 'index.html' : '' // add index.html if the path ends with a slash
+    )
+    const found = await fileExists(filePath)
+
```

Comment:
nit (non-blocking): Could add this to prevent path traversal, but not 100% necessary as this feature will be used in a controlled dev environment.

```suggestion
// Verify path is within the intended directory
const outputDirPath = path.resolve(staticPath, build.outputDirectory)
if (!filePath.startsWith(outputDirPath)) {
  return { found: false }
}
```

Reply:
Sounds good. I added two checks:
- the request url is matching the `publicPath` of the current build in the iteration
- the file path is a direct descendant of the `outputDirectory` of the build.